### PR TITLE
Enable auto refresh & stay logged in feature

### PR DESCRIPTION
**Pull request by larsschumachenko**
Hi,

I'd like to add 2 features:

- The functionality to use the --user-data-dir option to persist the browser data.
This enables the usage of cookies to skip the authentication as long as the cookie is valid.
As a result, credential refresh should be much smoother. This is also a requirement for the second feature I'd like to add.
- The second feature I'd like to add is the possibility to refresh all profiles with only one command. This will allow you to stay logged in while working with AWS, even if the maximum allowed session lifetime is configured to a low value on AWS side (eg. due to security constraints within a company).

Let me know your thoughts.

Best regards
Lars

### CODE REVIEW

**Comment by larsschumachenko** [comment]
Hi

I rebased my changes on version 1.7.0
There is one conflicting change, it's about adding the expiring time to the credentials file.
In alignment to the naming approach of the other variables (check here: https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html#API_AssumeRole_Examples), I've choose aws_expiration and a simple string presentation.
However, pull request #46 introduced the same data as json, with a different key: aws_session_expiration.

@dtjohnson I'd like to have the naming and data presentation aligned, but it's your call.

Happy to get some feedback on this pull request.

Best regards
Lars

**Comment by dtjohnson** [comment]
Thanks for the PR! I thought about session refreshes for a while but stopped when AWS allowed longer session times. I can appreciate a company wanting lower limits though.

Your solution is clever--both the repeated Azure logins and using cron for the refresh. I hadn't thought of those. My only concern is how flakey this is. The Puppeteer+Azure solution is not the most robust and repeatedly calling it in the background is bound to fail. But if folks want it as an option I'll include it.

I just had one alternative suggestion. Would your company allow you to add `sts:AssumeRole` to your SSO role's policy (assuming it doesn't have it)? If you add a statement like this to assume your own role (with your role ARN):
```json
{
    "Effect": "Allow",
    "Action": "sts:AssumeRole",
    "Resource": "arn:aws:iam::123456789012:role/your-role-name"
}
```
then you can repeatedly refresh your session just with the AWS API.

**Comment by larsschumachenko** [comment]
Thank you for your feedback.

I was thinking about this solution and came to the conclusion it would introduce a security issue.
If I would allow the role to assume itself and offboard a team member, this member could have still access if running a cronjob in the background.
I couldn't find a secure way of achieving a session refresh without going through azure ad.

Did I probably overlook something?

I'm aware that the Puppeteer+Azure solution is not the most robust one. The nice thing about using a stored session is skipping the whole login procedure. So the browser only need to follow a couple of redirects.
As long as the cookie is valid, and the team member is authorised to access AWS, it works quite ok.

**Comment by larsschumachenko** [comment]
Hi,

What do you think regarding my comments?
Can it be merged or will it be merged?

Merry Christmas
Lars

**Comment by TheBrigandier** [comment]
Would love to see this PR get approved, aws-azure-login would become so much more usable.

Thanks!

**Comment by larsschumachenko** [comment]
Hi,

Rebased my changes on current master and added the latest features to the login all mode as well.
Looking forward to get it merged, or to get some feedback.

Best regards
Lars

**Comment by TheBrigandier** [comment]
@mahdi-hosseini - any chance of getting this merged? Would love to see this functionality.

**Comment by larsschumachenko** [comment]
Nice to see this project active again, thank you 👍

Updated this PR to reflect latest changes and to fix some flaws in the flow when a user has chosen to stay logged in.

Summary of the changes:
- Added functionality to remember the cookies (rememberMe, using --user-data-dir of chromium)
- Added ability refresh all profiles without getting asked for input. Combine it with rememberMe and one can easily refresh all credentials with a cronjob
- As running the refresh with a cron leads to lot's of unnecessary refreshes a check for freshness and skip of refresh has been added as well.

Let me know your thoughts.

**Comment by larsschumachenko** [comment]
@kennethaasan @olemarkus 
Do you have any feedback for this MR? Would be happy to discuss it. Also happy if it gets accepted 🙂

**Comment by olemarkus** [comment]
Sorry for taking so long to get back to this. @kennethaasan and I will review the code the next few days. We both think this is a feature we want

File: lib/configureProfileAsync.js
Comment (kennethaasan) [inline lib/configureProfileAsync.js]: You need to add more protection here. I got an error trying to re-configure my existing profile. `default: (profile && profile.azure_default_remember_me && profile.azure_default_remember_me.toString()) || 'false',` should do.

File: bin/index.js
Comment (kennethaasan) [inline bin/index.js]: Looks like this can be compressed into one Promise.resolve. Something like this

```js
Promise.resolve()
    .then(() => {
        if (commander.allProfiles) {
            return login.loginAll(mode, disableSandbox, noPrompt, enableChromeNetworkService, awsNoVerifySsl, enableChromeSeamlessSso, forceRefresh);
        }

        if (commander.configure) return configureProfileAsync(profileName);
        return login.loginAsync(profileName, mode, disableSandbox, noPrompt, enableChromeNetworkService, awsNoVerifySsl, enableChromeSeamlessSso);
    })
    .catch(err => {
        if (err.name === "CLIError") {
            console.error(err.message);
            process.exit(2);
        } else {
            console.log(err);
        }
    });
```

File: lib/awsConfig.js
Comment (kennethaasan) [inline lib/awsConfig.js]: Not a big one, but I would prefer it if `refreshLimitInMs` was moved out next to `refreshLimit` because it's a constant

File: lib/login.js
Comment (kennethaasan) [inline lib/login.js]: Looks like this and the paths in `awsConfig.js` should be moved into a `paths.js` file maybe?

**Review by kennethaasan — CHANGES_REQUESTED** [review]
Great contribution, thank you! 
I made a few comments, and once those are fixed we can merge.

**Comment by larsschumachenko** [comment]
I guess I addressed all topics, looks indeed nicer now.

**Review by kennethaasan — APPROVED** [review]

**Pull request merged**
